payjoin/src/uri.rs

```diff
@@ -1,12 +1,109 @@
-use bitcoin::Network;
+use bip21::{DeserializationError, DeserializeParams};
+use bitcoin::address::{Error, NetworkChecked, NetworkUnchecked, NetworkValidation};
+use bitcoin::{Address, Amount, Network};
 use url::Url;
 
-/// Payjoin enum represents payjoin parameters in
-/// different states.
+/// Payjoin Uri represents a bip21 uri with additional
+/// payjoin parameters.
+pub struct PayjoinUri<
+    'a,
+    N: NetworkValidation,
+    P: DeserializeParams<'a> + DeserializationError + Sized,
+> {
+    inner: bip21::Uri<'a, N, P>,
+}
+
+impl<'a, N: NetworkValidation, P: DeserializeParams<'a> + DeserializationError + Sized>
+    From<bip21::Uri<'a, N, P>> for PayjoinUri<'a, N, P>
+{
+    fn from(value: bip21::Uri<'a, N, P>) -> Self { Self { inner: value } }
+}
+
+impl<'a> PayjoinUri<'a, NetworkUnchecked, Payjoin> {
+    /// Marks network of this address as checked.
+    pub fn assume_checked(self) -> PayjoinUri<'a, NetworkChecked, Payjoin> {
+        let checked_address = self.inner.address.assume_checked();
+        let mut uri = bip21::Uri::with_extras(checked_address, self.inner.extras);
+        uri.amount = self.inner.amount;
+        uri.label = self.inner.label;
+        uri.message = self.inner.message;
+        uri.into()
+    }
+    /// Checks whether network of this address is as required.
+    pub fn require_network(
+        self,
+        network: Network,
+    ) -> Result<PayjoinUri<'a, NetworkChecked, Payjoin>, Error> {
+        let checked_address = self.inner.address.require_network(network)?;
+        let mut uri = bip21::Uri::with_extras(checked_address, self.inner.extras);
+        uri.amount = self.inner.amount;
+        uri.label = self.inner.label;
+        uri.message = self.inner.message;
+        Ok(uri.into())
+    }
+}
+
+impl<'a, N: NetworkValidation> PayjoinUri<'a, N, Payjoin> {
+    /// Returns the receiver bitcoin address from this uri.
+    pub fn address(&self) -> &Address<N> { &self.inner.address }
+    /// Returns the amount in bitcoin from this uri.
+    pub fn amount(&self) -> Option<Amount> { self.inner.amount }
+    /// Returns the endpoint url from this uri.
+    pub fn endpoint(&self) -> Option<&Url> { self.inner.extras.endpoint() }
+    /// Returns whether output substitution is disabled for this uri.
+    pub fn disable_output_substitution(&self) -> bool {
+        self.inner.extras.disable_output_substitution()
+    }
+    /// Returns whether payjoin is supported for this uri.
+    pub fn pj_is_supported(&self) -> bool { self.inner.extras.pj_is_supported() }
+    #[cfg(feature = "v2")]
+    /// Returns the ohttp config for this uri.
+    pub fn ohttp_config(&self) -> Option<&ohttp::KeyConfig> { self.inner.extras.ohttp_config() }
+}
+
+impl<'a> PayjoinUri<'a, NetworkChecked, Payjoin> {
+    fn with_extras(address: Address<NetworkChecked>, extras: Payjoin) -> Self {
+        Self { inner: bip21::Uri::with_extras(address, extras) }
+    }
+    pub fn label(&self) -> Option<String> {
```

Comment:
should be moved to `impl<'a, N: NetworkValidation> PayjoinUri<'a, N, Payjoin> {`